File: backend/scorer.py
# =============================================================================
# ZenNode Backend — Cognitive Load Scorer (Layer A)
#
# This is the deterministic math engine. It takes a BehavioralSnapshot,
# normalizes the raw metrics, computes a weighted cognitive load score (0–100),
# and classifies the developer's cognitive state.
#
# This runs on EVERY snapshot (every 30s). It's fast, pure math, no LLM.
# =============================================================================

from models import BehavioralSnapshot, CognitiveReport, CognitiveState, MetricBreakdown
from thresholds import (
    WEIGHT_SWITCH_RATE, WEIGHT_ERROR_RATE, WEIGHT_UNDO_RATE,
    WEIGHT_IDLE_RATIO, WEIGHT_PASTE_RATIO,
    SWITCH_RATE_CAP, ERROR_RATE_CAP, UNDO_RATE_CAP,
    IDLE_RATIO_CAP, PASTE_RATIO_CAP,
    THRESHOLD_FLOW_MAX, THRESHOLD_FRICTION_MAX, THRESHOLD_FATIGUE_MAX,
    THRESHOLD_THEME_SHIFT, MIN_KEYSTROKES_FOR_SCORING, EMA_ALPHA,
)


def _clamp(value: float, lo: float = 0.0, hi: float = 100.0) -> float:
    """Clamp a value between lo and hi."""
    return max(lo, min(hi, value))


def _normalize(raw: float, cap: float) -> float:
    """Normalize a raw value to 0–100 scale using the cap as the ceiling."""
    if cap <= 0:
        return 0.0
    return _clamp((raw / cap) * 100.0)


def compute_metrics(snapshot: BehavioralSnapshot) -> MetricBreakdown:
    """
    Convert raw behavioral counts into normalized 0–100 metric scores.

    Each metric represents a different dimension of cognitive stress:
      - switchRate:  context switching intensity
      - errorRate:   error correction intensity
      - undoRate:    decision reversal intensity
      - idleRatio:   disengagement level
      - pasteRatio:  passive code acceptance level
    """
    duration_min = snapshot.duration_ms / 60_000.0
    if duration_min <= 0:
        duration_min = 0.5  # safety floor: assume at least 30s

    # ── SwitchRate: tab switches per minute ──
    raw_switch_rate = snapshot.tab_switches / duration_min
    switch_rate = _normalize(raw_switch_rate, SWITCH_RATE_CAP)

    # ── ErrorRate: backspaces as % of total keystrokes ──
    total_key_actions = snapshot.keystrokes
    if total_key_actions > 0:
        raw_error_rate = (snapshot.backspaces / total_key_actions) * 100.0
    else:
        raw_error_rate = 0.0
    error_rate = _normalize(raw_error_rate, ERROR_RATE_CAP)

    # ── UndoRate: undos per minute ──
    raw_undo_rate = snapshot.undos / duration_min
    undo_rate = _normalize(raw_undo_rate, UNDO_RATE_CAP)

    # ── IdleRatio: idle ms as % of total duration ──
    raw_idle_ratio = (snapshot.idle_ms / snapshot.duration_ms) * 100.0
    idle_ratio = _normalize(raw_idle_ratio, IDLE_RATIO_CAP)

    # ── PasteRatio: pasted chars as % of total chars ──
    if snapshot.total_chars > 0:
        raw_paste_ratio = (snapshot.pasted_chars / snapshot.total_chars) * 100.0
    else:
        raw_paste_ratio = 0.0
    paste_ratio = _normalize(raw_paste_ratio, PASTE_RATIO_CAP)

    return MetricBreakdown(
        switch_rate=round(switch_rate, 1),
        error_rate=round(error_rate, 1),
        undo_rate=round(undo_rate, 1),
        idle_ratio=round(idle_ratio, 1),
        paste_ratio=round(paste_ratio, 1),
    )


def compute_score(metrics: MetricBreakdown) -> float:
    """
    Compute the weighted cognitive load score (0–100).

    S = w1·SwitchRate + w2·ErrorRate + w3·UndoRate + w4·IdleRatio + w5·PasteRatio
    """
    raw_score = (
        WEIGHT_SWITCH_RATE * metrics.switch_rate +
        WEIGHT_ERROR_RATE  * metrics.error_rate +
        WEIGHT_UNDO_RATE   * metrics.undo_rate +
        WEIGHT_IDLE_RATIO  * metrics.idle_ratio +
        WEIGHT_PASTE_RATIO * metrics.paste_ratio
    )
    return _clamp(round(raw_score, 1))


def classify_state(score: float) -> CognitiveState:
    """
    Map a cognitive load score to one of the four states.

      0–30:  Flow      (🟢 green)
      31–60: Friction  (🟡 yellow)
      61–80: Fatigue   (🟠 orange)
      81–100: Overload (🔴 red)
    """
    if score <= THRESHOLD_FLOW_MAX:
        return CognitiveState.FLOW
    elif score <= THRESHOLD_FRICTION_MAX:
        return CognitiveState.FRICTION
    elif score <= THRESHOLD_FATIGUE_MAX:
        return CognitiveState.FATIGUE
    else:
        return CognitiveState.OVERLOAD


def should_shift_theme(score: float) -> bool:
    """Whether the score warrants a warm-amber theme shift."""
    return score >= THRESHOLD_THEME_SHIFT


class CognitiveScorer:
    """
    Stateful scorer that maintains an exponential moving average (EMA)
    across snapshots for smooth, non-jittery score transitions.

    Usage:
        scorer = CognitiveScorer()
        report = scorer.score(snapshot)
    """

    def __init__(self) -> None:
        self._ema_score: float | None = None

    def score(self, snapshot: BehavioralSnapshot) -> CognitiveReport:
        """
        Score a behavioral snapshot and return a full cognitive report.

        If the developer is mostly inactive (< MIN_KEYSTROKES_FOR_SCORING),
        we return a neutral "flow" state instead of scoring on noise.
        """
        # ── Handle low-activity snapshots gracefully ──
        if snapshot.keystrokes < MIN_KEYSTROKES_FOR_SCORING:
            return self._neutral_report()

        # ── Layer A: Compute normalized metrics ──
        metrics = compute_metrics(snapshot)

        # ── Compute raw score ──
        raw_score = compute_score(metrics)

        # ── Apply EMA smoothing ──
        if self._ema_score is None:
            smoothed_score = raw_score
        else:
            smoothed_score = round(
                EMA_ALPHA * raw_score + (1 - EMA_ALPHA) * self._ema_score, 1
            )
        self._ema_score = smoothed_score

        # ── Classify state ──
        state = classify_state(smoothed_score)
        theme_shift = should_shift_theme(smoothed_score)

        return CognitiveReport(
            score=smoothed_score,
            state=state,
            theme_shift=theme_shift,
            intervention=None,  # LLM fills this in later (Feature #9)
            metrics=metrics,
        )

    def reset(self) -> None:
        """Reset the EMA history (e.g., on session reset)."""
        self._ema_score = None

    @property
    def last_score(self) -> float | None:
        """The most recent smoothed score, or None if no snapshots scored yet."""
        return self._ema_score

    def _neutral_report(self) -> CognitiveReport:
        """Return a calm, neutral report for low-activity periods."""
        neutral_metrics = MetricBreakdown(
            switch_rate=0, error_rate=0, undo_rate=0,
            idle_ratio=0, paste_ratio=0,
        )
        # Gently decay the EMA toward 0 during idle periods
        if self._ema_score is not None:
            self._ema_score = round(self._ema_score * 0.8, 1)

        return CognitiveReport(
            score=self._ema_score or 0.0,
            state=classify_state(self._ema_score or 0.0),
            theme_shift=False,
            intervention=None,
            metrics=neutral_metrics,
        )


def reset_scorer(scorer: CognitiveScorer) -> None:
    """Convenience function to reset the scorer from outside."""
    scorer.reset()

def get_last_score(scorer: CognitiveScorer) -> float | None:
    """Convenience function to get the last smoothed score from outside."""
    return scorer.last_score

File: src/scorer.ts
// ============================================================================
// ZenNode — Cognitive Load Scorer (TypeScript port of scorer.py)
//
// Pure math engine. Takes a BehavioralSnapshot, normalizes metrics,
// computes a weighted cognitive load score (0–100), and classifies state.
// Runs entirely inside the VS Code extension — no server needed.
// ============================================================================

import { BehavioralSnapshot, CognitiveReport, CognitiveState, MetricBreakdown } from './types';

// ── Scoring weights (must sum to 1.0) ──────────────────────────────────────
const WEIGHT_SWITCH_RATE = 0.25;
const WEIGHT_ERROR_RATE = 0.20;
const WEIGHT_UNDO_RATE = 0.25;
const WEIGHT_IDLE_RATIO = 0.15;
const WEIGHT_PASTE_RATIO = 0.15;

// ── Normalization caps ─────────────────────────────────────────────────────
const SWITCH_RATE_CAP = 15.0;   // switches/minute → 100
const ERROR_RATE_CAP = 50.0;   // backspace % → 100
const UNDO_RATE_CAP = 10.0;   // undos/minute → 100
const IDLE_RATIO_CAP = 80.0;   // idle % → 100
const PASTE_RATIO_CAP = 90.0;   // paste % → 100

// ── State thresholds ───────────────────────────────────────────────────────
// export const THRESHOLD_FLOW_MAX = 10;  // TEST: lowered from 30
// export const THRESHOLD_FRICTION_MAX = 20;  // TEST: lowered from 60
// export const THRESHOLD_FATIGUE_MAX = 30;  // TEST: lowered from 80
// export const THRESHOLD_THEME_SHIFT = 30;  // TEST: lowered from 80
// export const THRESHOLD_LLM_TRIGGER = 30;  // TEST: lowered from 80

export const THRESHOLD_FLOW_MAX = 30;
export const THRESHOLD_FRICTION_MAX = 60;
export const THRESHOLD_FATIGUE_MAX = 80;
export const THRESHOLD_THEME_SHIFT = 80;
export const THRESHOLD_LLM_TRIGGER = 80;

// ── EMA smoothing ──────────────────────────────────────────────────────────
const EMA_ALPHA = 0.6;
const MIN_KEYSTROKES = 3;

// ── Helpers ────────────────────────────────────────────────────────────────

function clamp(value: number, lo = 0, hi = 100): number {
    return Math.max(lo, Math.min(hi, value));
}

function normalize(raw: number, cap: number): number {
    if (cap <= 0) { return 0; }
    return clamp((raw / cap) * 100);
}

function round1(n: number): number {
    return Math.round(n * 10) / 10;
}

// ── Metric computation ─────────────────────────────────────────────────────

function computeMetrics(snapshot: BehavioralSnapshot): MetricBreakdown {
    const durationMin = Math.max(snapshot.durationMs / 60_000, 0.5);

    const rawSwitchRate = snapshot.tabSwitches / durationMin;
    const switchRate = normalize(rawSwitchRate, SWITCH_RATE_CAP);

    const rawErrorRate = snapshot.keystrokes > 0
        ? (snapshot.backspaces / snapshot.keystrokes) * 100
        : 0;
    const errorRate = normalize(rawErrorRate, ERROR_RATE_CAP);

    const rawUndoRate = snapshot.undos / durationMin;
    const undoRate = normalize(rawUndoRate, UNDO_RATE_CAP);

    const rawIdleRatio = (snapshot.idleMs / snapshot.durationMs) * 100;
    const idleRatio = normalize(rawIdleRatio, IDLE_RATIO_CAP);

    const rawPasteRatio = snapshot.totalChars > 0
        ? (snapshot.pastedChars / snapshot.totalChars) * 100
        : 0;
    const pasteRatio = normalize(rawPasteRatio, PASTE_RATIO_CAP);

    return {
        switchRate: round1(switchRate),
        errorRate: round1(errorRate),
        undoRate: round1(undoRate),
        idleRatio: round1(idleRatio),
        pasteRatio: round1(pasteRatio),
    };
}

function computeScore(metrics: MetricBreakdown): number {
    const raw =
        WEIGHT_SWITCH_RATE * metrics.switchRate +
        WEIGHT_ERROR_RATE * metrics.errorRate +
        WEIGHT_UNDO_RATE * metrics.undoRate +
        WEIGHT_IDLE_RATIO * metrics.idleRatio +
        WEIGHT_PASTE_RATIO * metrics.pasteRatio;
    return clamp(round1(raw));
}

function classifyState(score: number): CognitiveState {
    if (score <= THRESHOLD_FLOW_MAX) { return 'flow'; }
    if (score <= THRESHOLD_FRICTION_MAX) { return 'friction'; }
    if (score <= THRESHOLD_FATIGUE_MAX) { return 'fatigue'; }
    return 'overload';
}

// ── Stateful scorer with EMA ───────────────────────────────────────────────

export class CognitiveScorer {
    private _emaScore: number | null = null;

    score(snapshot: BehavioralSnapshot): CognitiveReport {
        if (snapshot.keystrokes < MIN_KEYSTROKES) {
            return this._neutralReport();
        }

        const metrics = computeMetrics(snapshot);
        const rawScore = computeScore(metrics);

        const smoothed = this._emaScore === null
            ? rawScore
            : round1(EMA_ALPHA * rawScore + (1 - EMA_ALPHA) * this._emaScore);

        this._emaScore = smoothed;

        return {
            score: smoothed,
            state: classifyState(smoothed),
            themeShift: smoothed >= THRESHOLD_THEME_SHIFT,
            intervention: null,
            metrics,
        };
    }

    reset(): void {
        this._emaScore = null;
    }

    get lastScore(): number | null {
        return this._emaScore;
    }

    /** Restore EMA from a persisted value (e.g. after VS Code restart). */
    restoreEma(score: number): void {
        this._emaScore = score;
    }

    private _neutralReport(): CognitiveReport {
        // Gently decay EMA toward 0 during idle periods
        if (this._emaScore !== null) {
            this._emaScore = round1(this._emaScore * 0.8);
        }
        const score = this._emaScore ?? 0;
        return {
            score,
            state: classifyState(score),
            themeShift: false,
            intervention: null,
            metrics: { switchRate: 0, errorRate: 0, undoRate: 0, idleRatio: 0, pasteRatio: 0 },
        };
    }
}
